src/animalShelter/AnimalShelter.ts

```diff
@@ -28,104 +26,52 @@ export default class AnimalShelter {
-  }
-  getReptileLimit() {
-    return this._reptileLimit;
-  }
-
   showAnimals() {
+
     console.log(this.animals);
-    console.log(`lions: ${this.lions}`);
-    console.log(`crocodiles: ${this.crocodiles}`);
-    console.log(`snakes: ${this.snakes}`);
-    console.log(`tortoise: ${this.tortoise}`);
+    console.log("mammals: " + this.getMammalsCount());
+    console.log("reptiles: " + this.getReptileCount());
+    console.log(`lions: ${this.animals.filter(x => x instanceof Lion).length}`);
+    console.log(`crocodiles: ${this.animals.filter(x => x instanceof Crocodile).length}`);
+    console.log(`squirrels: ${this.animals.filter(x => x instanceof Squirrel).length}`);
+    console.log(`snakes: ${this.animals.filter(x => x instanceof Snake).length}`);
+    console.log(`tortoise: ${this.animals.filter(x => x instanceof Tortoise).length}`);
   }
 
   addAnimal(animal: Animal): void {
-    if (animal instanceof Mammal) {
-      if (this.mammalsNumber >= this._mammalLimit) {
-        console.log(
-          "Sorry, there is no space for another mammal in the Animal Shelter!"
-        );
-        return;
-      }
 
-      if (animal instanceof Lion) {
-        if (animal.getHome() !== undefined) {
-          const family = families.find(
-            (pack) => pack.name === animal.getHome()
-          );
-          family?.removeAnimal(animal as Lion);
-          animal.setHome(undefined);
-        }
-        this.lions++;
-      } else if (animal instanceof Squirrel) {
-        this.squirrels++;
-      }
-      this.animals.push(animal);
-      animal.setHome("Animal Shelter");
-      console.log(`${animal.getName()} has been added to the Animal Shelter`);
-    } else if (animal) {
-      if (this.reptileNumbers >= this._reptileLimit) {
-        console.log(
-          "Sorry, there is no space for another reptile in the Animal Shelter!"
-        );
-        return;
-      }
-      if (animal instanceof Crocodile) {
-        if (animal.getHome() !== undefined) {
-          const pack = families.find((pack) => pack.name === animal.getHome());
-          pack?.removeAnimal(animal);
-          animal.setHome(undefined);
-        }
-        this.crocodiles++;
-      } else if (animal instanceof Tortoise) {
-        this.tortoise++;
-      } else if (animal instanceof Snake) {
-        this.snakes++;
+    if (this.getReptileCount() + 1 > this.reptileLimit) {
+      console.error(
+        "Sorry, there is no space for another reptile in the Animal Shelter!"
+      );
+      return;
+    }
+    else if (this.getMammalsCount() + 1 > this.mammalLimit) {
+      console.error(
+        "Sorry, there is no space for another mammal in the Animal Shelter!"
+      );
+      return;
+    }
+    let isRemovalValid: boolean = true;
+    if (animal.getCanHaveFamily()) {
+      const family = families.find(
+        (pack) => pack.name === animal.getHome()
+      );
+      if (family) {
+        isRemovalValid = family.removeAnimal(animal);
       }
+    }
```

Comment:
I think following this logic, if the animal isn't already in a home, it cannot be removed, hence it cannot be added to the shelter

Reply:
I implemented this logic
 if (families.find(pack => pack.name === animal.getHome()) !== undefined){
      families.find(pack => pack.name === animal.getHome())?.removeAnimal(animal);
}
this.animals.push(animal);
animal.setHome("Animal Shelter");
console.log(`${animal.getName()} has been added to the Animal Shelter`);

It is basically the same, just the addition of the animal to the shelter isn't dependent on its removal.
Is it better to save families.find(pack => pack.name === animal.getHome()) into a constant or keep it this way. In my opinion making it a constant saves us from second time checking if its present or undefined

Reply:
const family = families.find(pack => pack.name === animal.getHome());
    if (family !== undefined){
      family.removeAnimal(animal);
    } 
    this.animals.push(animal);
    animal.setHome("Animal Shelter");
    console.log(`${animal.getName()} has been added to the Animal Shelter`);
if the family is present, family.removeAnimal(), then push and other operations.
If family is not present, push and other operations.

Reply:
if (families.find(pack => pack.name === animal.getHome())?.removeAnimal(animal)){
      this.animals.push(animal);
      animal.setHome("Animal Shelter");
      console.log(`${animal.getName()} has been added to the Animal Shelter`);
    } 
    else {
      this.animals.push(animal);
      animal.setHome("Animal Shelter");
      console.log(`${animal.getName()} has been added to the Animal Shelter`);
    }

Reply:
is it ok if we leave the body of the if statement empty, remove the else and just place the push... code afterwards?

Reply:
it is ok, you can also try to invert the whole if  and then your else becomes the if .   
This part is not ok 
 remove the else and just place the push... code afterwards   because it will always execute the  push...  no matter the result from the if statement